fairseq/sequence_generator.py

```diff
@@ -366,8 +368,13 @@ def calculate_banned_tokens(bbsz_idx):
                     cand_indices = prefix_tokens[:, step].view(-1, 1).repeat(1, cand_size)
                     cand_beams = torch.zeros_like(cand_indices)
 
-                    # handle prefixes of different lengths
-                    partial_prefix_mask = prefix_tokens[:, step].eq(self.pad)
+                # handle prefixes of different lengths
+                # when step == prefix_tokens.size(1), we'll have new free-decoding batches
+                if prefix_tokens is not None and step <= prefix_tokens.size(1):
```

Comment:
There seems to be one more minor corner case to be covered.

1) prefix_tokens could have 0 prefixes. In this case we will enter if condition above because `step <= prefix_tokens.size(1)` where step is 0 and prefix_tokens.size(1) is 0 and `cand_scores` wasn't initialized. We will get a crash. To repro this case we can enter interactive mode and press <enter> without entering any text:
```
~/fairseq-py$ CUDA_VISIBLE_DEVICES=1 python fairseq_cli/interactive.py ~/downloads/newwiki/private/home/abaevski/data/wiki103/ --path ~/downloads/newwiki/wiki103.pt --task='language_modeling' --nbest 5
...
| Type the input sentence and press return:

Traceback (most recent call last):
  File "fairseq_cli/interactive.py", line 173, in <module>
    cli_main()
  File "fairseq_cli/interactive.py", line 169, in cli_main
    main(args)
  File "fairseq_cli/interactive.py", line 130, in main
    translations = task.inference_step(generator, models, sample)
  File "/private/home/oxo/fairseq-py/fairseq/tasks/language_modeling.py", line 229, in inference_step
    return generator.generate(models, sample, prefix_tokens=prefix_tokens)
  File "/private/home/oxo/.conda/envs/fairseq-20190509/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 43, in decorate_no_grad
    return func(*args, **kwargs)
  File "/private/home/oxo/fairseq-py/fairseq/sequence_generator.py", line 387, in generate
    cand_scores[partial_prefix_mask] = partial_scores[partial_prefix_mask]
UnboundLocalError: local variable 'cand_scores' referenced before assignment
```

Reply:
Hi @okhonko , thanks for your review. It seems to me not a bug in sequence_generate.py, but one for passing a no-element prefix_tokens. I think a no-element prefix_token should be set to None.

Reply:
@yilinyang7 , yes, this seems to be the case. You can fix it in this pull request or I can send a quick fix later so that we will not introduce new crashes with this code change.